extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/secret/KsmClient.java

```diff
@@ -250,13 +292,38 @@ private void validateCache() throws GuacamoleException {
                 String hostname = recordService.getHostname(record);
                 addRecordForHost(record, hostname);
 
-                // Store based on username ONLY if no hostname (will otherwise
-                // result in ambiguous entries for servers tied to identical
-                // accounts)
-                if (hostname == null)
-                    addRecordForLogin(record, recordService.getUsername(record));
+                // ... and domain
+                String domain = recordService.getDomain(record);
+                addRecordForDomain(record, domain);
+
+                // Get the username off of the record
+                String username = recordService.getUsername(record);
+
+                // If we have a username, and there isn't already a domain explicitly defined
+                if (username != null && domain == null
+                        && confService.getSplitWindowsUsernames()) {
 
-            });
+                    // Attempt to split out the domain of the username
+                    WindowsUsername usernameAndDomain = (
+                            WindowsUsername.splitWindowsUsernameFromDomain(username));
+
+                    // Use the username-split domain if not already set explicitly
+                    if (usernameAndDomain.hasDomain())
+                        domain = usernameAndDomain.getDomain();
+                        addRecordForDomain(record, domain);
+
+                }
+
+                // If domain matching is not enabled for user records,
+                // explicitly set all domains to null to allow matching
+                // on username only
+                if (!confService.getMatchUserRecordsByDomain())
+                    domain = null;
+
+                // Store the login by username and domain
+                addRecordForLogin(record, username, domain);
```

Comment:
Given a record with a login like `DOMAIN\Username`, no field named "domain", and with domain splitting and domain matching enabled, this looks like it will:

1. Index that record as the username `DOMAIN\Username`.
2. Index that record again with the domain+username pair.

Is that intended?

My thinking was that the intended behavior was:

* Domain (if present) and username are pulled from the record. If splitting is enabled, that domain _might_ come from splitting the username, in which case the username will be missing the domain.
* Records are indexed by domain (however that gets retrieved) and by either domain+username (if domain matching is enabled) or just username (if domain matching is not enabled).

If the above is true, then I'd think it would be better to abstract this away beneath the record service, such that `getUsername()` and `getDomain()` retrieve what we are considering to be the record's username and domain, transparently taking splitting into account.

Reply:
Where do you see it indexing the record twice, with `DOMAIN\Username` and domain+username? As far as I can tell, the code is already doing what you're suggesting above.

Reply:
Where did you see it indexing by both the raw username, and domain+username? I only see a single call to `addRecordForLogin`, which should do either one of those depending on `getSplitWindowsUsernames()`.

There was a bug here where I didn't correctly remove the username from the domain if splitting is enabled, but that should be fixed now.

Everything should be working the way you suggest above.

Reply:
You're right - I think I must have imagined an `addRecordForLogin()` call below `recordService.getUsername()`, but there isn't actually one there. It's just retrieving the username.

> Everything should be working the way you suggest above.

Do you think it's not worth moving the domain/username retrieval logic into the record service? I still find it a bit odd that record service abstracts away the details of retrieval for everything _but_ the splitting case, where the caller (in this case `KsmClient` and `KsmSecretService`) must always manually check and split.

Reply:
Sure, how does this look?

Reply:
Muuuuuuuuch better.